ilastik/workflows/trainableDomainAdaptation/_localTrainableDomainAdaptationWorkflow.py

```diff
@@ -0,0 +1,205 @@
+    Furthermore, the ilastik app will try to find a tiktorch version
+    (usually the case when bundled for distribution) if the parameter is not
+    supplied. If it cannot be found, workflow is not available.
+    """
+
+    auto_register = True
+    workflowName = "Trainable Domain Adaptation (Local) (beta)"
+    workflowDescription = "Allows to apply bioimage.io shallow2deep models on your data using bundled tiktorch"
+
+    @property
+    def ExportNames(self):
+        @enum.unique
+        class ExportNames(SlotNameEnum):
+            PROBABILITIES = enum.auto()
+            LABELS = enum.auto()
+            # TODOS: add NN output, too
+
+        return ExportNames
+
+    def __init__(self, shell, headless, workflow_cmdline_args, project_creation_args, *args, **kwargs):
+        tiktorch_exe_path = runtime_cfg.tiktorch_executable
+        if not tiktorch_exe_path:
+            raise RuntimeError("No tiktorch-executable specified")
+
+        self._launcher = LocalServerLauncher(tiktorch_exe_path)
+        super().__init__(shell, headless, workflow_cmdline_args, project_creation_args, *args, **kwargs)
+
+    def _createFeatureSelectionApplet(self):
+        return FeatureSelectionApplet(self, "Feature Selection", "FeatureSelections")
+
+    def _create_local_connection(self):
+        conn_str = self._launcher.start()
+        return conn_str
+
+    def _createClassifierApplet(self, headless=False, conn_str=None):
+        if not headless or not conn_str:
+            conn_str = self._create_local_connection()
+
+        self.featureSelectionApplet = self._createFeatureSelectionApplet()
+        self.applets.append(self.featureSelectionApplet)
+        srv_config = ServerConfig(id="auto", address=conn_str, devices=[Device(id="cpu", name="cpu", enabled=True)])
+        connFactory = tiktorch.TiktorchConnectionFactory()
+        conn = connFactory.ensure_connection(srv_config)
+
+        devices = conn.get_devices()
+        preferred_cuda_device_id = runtime_cfg.preferred_cuda_device_id
+        device_ids = [dev[0] for dev in devices]
+        cuda_devices = tuple(d for d in device_ids if d.startswith("cuda"))
+
+        if preferred_cuda_device_id not in device_ids:
+            if preferred_cuda_device_id:
+                logger.warning(f"Could nor find preferred cuda device {preferred_cuda_device_id}")
+            try:
+                preferred_cuda_device_id = cuda_devices[0]
+            except IndexError:
+                preferred_cuda_device_id = "cpu"
+
+            logger.info(f"Using default device for Neural Network Workflow {preferred_cuda_device_id}")
+        else:
+            logger.info(f"Using specified device for Neural Network Workflow {preferred_cuda_device_id}")
+
+        device_name = devices[device_ids.index(preferred_cuda_device_id)][1]
+
+        srv_config = srv_config.evolve(devices=[Device(id=preferred_cuda_device_id, name=device_name, enabled=True)])
```

Comment:
The name `preferred_cuda_device_id` is somewhat misleading because it could be `"cpu"` as well:
```suggestion
        devices = {d[0]: d[1] for d in conn.get_devices()}
        device_id = runtime_cfg.preferred_cuda_device_id

        if device_id in devices:
            logger.info(f"Neural Network Workflow: using preferred device {device_id}")
        else:
            if device_id:
                logger.warning(f"Neural Network Workflow: preferred device {device_id} not found")
            device_id = next((k for k in devices if k.startswith("cuda")), "cpu")
            logger.info(f"Neural Network Workflow: using default device {device_id}")
        
        srv_config = srv_config.evolve(devices=[Device(id=device_id, name=devices[device_id], enabled=True)])
```

Reply:
beautiful - thanks. An yeah I also remember our discussion about the command line flag which then changed from `--cuda-device` to `--nn-device` :)